### PR TITLE
Fixes implementation of With methods of recoverable trees

### PR DESCRIPTION
**Pull request by tmat**
Fixes https://github.com/dotnet/roslyn/issues/63834

### CODE REVIEW

**Comment by tmat** [comment]
@jasonmalinowski PTAL

File: src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs [hunk 1]
```diff
@@ -24,29 +24,33 @@ private partial class CSharpSyntaxTreeFactoryService
...
+            internal sealed class RecoverableSyntaxTree : CSharpSyntaxTree, IRecoverableSyntaxTree<CSharpSyntaxNode>, ICachedObjectOwner
             {
-                private readonly RecoverableSyntaxRoot<CompilationUnitSyntax> _recoverableRoot;
+                private readonly RecoverableSyntaxRoot<CSharpSyntaxNode> _recoverableRoot;
                 private readonly SyntaxTreeInfo _info;
-                private readonly IProjectCacheHostService _projectCacheService;
+                private readonly AbstractSyntaxTreeFactoryService _service;
                 private readonly ProjectId _cacheKey;
 
+                public override bool HasCompilationUnitRoot { get; }
```
Comment (genlu) [inline src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs:34]: What would be the effect of this change (i.e. allowing non CompilationUnit as root here)?

File: src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs [hunk 2]
```diff
@@ -28,7 +28,7 @@ internal sealed class RecoverableSyntaxTree : CSharpSyntaxTree, IRecoverableSynt
             {
                 private readonly RecoverableSyntaxRoot<CompilationUnitSyntax> _recoverableRoot;
                 private readonly SyntaxTreeInfo _info;
-                private readonly IProjectCacheHostService _projectCacheService;
+                private readonly AbstractSyntaxTreeFactoryService _service;
```
Comment (genlu) [inline src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs:31]: Just curious, why this change? It doesn't seem to make any difference to me.

File: src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs (hunk 1 above)
Comment (tmat) [reply src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs:34]: Not sure what you mean. `WithRootAndOptions` accepts any syntax node as the root, so if we change the root of a recoverable tree to something that's not CompilationUnit we get a recoverable tree with that root.

Comment (genlu) [reply src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs:34]: Sorry my question was quite vague. I'm wondering in what scenario this is needed, given that it doesn't seem to relate to the bug fix. Also consider covering this in the test

**Review by genlu — APPROVED** [review]

File: src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs (hunk 1 above)
Comment (tmat) [reply src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs:34]: Some features do this as I discovered when tests were failing. So it is covered by tests already :)

File: src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs [hunk 3]
```diff
@@ -143,19 +134,20 @@ public override SyntaxReference GetReference(SyntaxNode node)
...
 
                 public override SyntaxTree WithRootAndOptions(SyntaxNode root, ParseOptions options)
                 {
-                    if (ReferenceEquals(_info.Options, options) && this.TryGetRoot(out var oldRoot) && ReferenceEquals(root, oldRoot))
+                    if (ReferenceEquals(_info.Options, options) && TryGetRoot(out var oldRoot) && ReferenceEquals(root, oldRoot))
                     {
                         return this;
                     }
 
-                    return Create((CSharpSyntaxNode)root, this.Options, _info.FilePath);
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs:158]: So we just never created a recoverable tree further from here???

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs [hunk 4]
```diff
@@ -2425,6 +2425,69 @@ private static ObjectReference<Compilation> GetObservedCompilation(Solution solu
...
+        [WorkItem(63834, "https://github.com/dotnet/roslyn/issues/63834")]
+        public void RecoverableTree_With()
+        {
+            using var workspace = CreateWorkspaceWithRecoverableSyntaxTreesAndWeakCompilations();
+
+            var pid = ProjectId.CreateNewId();
+            var did = DocumentId.CreateNewId(pid);
+
+            var sol = workspace.CurrentSolution
+                .AddProject(pid, "goo", "goo.dll", LanguageNames.CSharp)
```
Comment (jasonmalinowski) [inline src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs:2438]: Can you switch this to a Theory to cover VB as well?

**Review by jasonmalinowski — APPROVED** [review]
I think is the most confused I've been on a PR for awhile, not because the change isn't good, but because the old code was totally strange.

@tmat does this need a similar change for VB?

File: src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs (hunk 3 above)
Comment (tmat) [reply src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs:158]: Right.

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs (hunk 4 above)
Comment (tmat) [reply src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs]: Done for VB.
